### ISSUE TITLE
Past due timers or "RunOnStartup = true" on v3 halts the HTTP server listener until startup functions have completed. 

### ISSUE THREAD

**Issue by rwsanders** [Azure/azure-functions-host#5529]
In upgrading our function app from v2 to v3 we incurred a breaking change when running on localhost where a function that that is set to run at startup using "RunOnStartup = true" was attempting to execute via REST another azure function within the same project.  This results in the following error "No connection could be made because the target machine actively refused it."

This type of orchestration works on v2 but not on v3 when executing on my local machine.  My temporary work around is to change "RunOnStartup = false" and manually trigger from our deployment pipeline.

**Comment by mathewc** [comment]
Yes, @paulbatum pointed this out as well a while ago, noting that the listener startup code [here](https://github.com/Azure/azure-webjobs-sdk-extensions/blob/3d7b2b2bc75927f3e84ed231f14c175ff342ef23/src/WebJobs.Extensions/Extensions/Timers/Listener/TimerListener.cs#L124) will block on an invocation, both in RunOnStartup cases as well as when past due.

Likely what we should do is make any startup path invocations async - fire them off in the background, and only start the normal schedule after that initial invocation is complete.

You're not using RunOnStartup=true in production are you? See the warning [here](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-timer?tabs=csharp#configuration) against that.

**Comment by rwsanders** [comment]
we are using RunOnStartup in prod, thanks for linking the warning.  We intend to deviate from this pattern as we were shooting ourselves in the foot by running many unrelated processes at startup which resulted in getting throttled on our azure resource manager calls.  it just so happened we still have some code using this parameter and noticed the issue when updating.

**Comment by rwsanders** [comment]
Another observation with this bug is that if I set my timer to 1 minute or another low amount, then Azure Functions will opt to run at start (before the listener is invoked) even with RunOnStartup = false just because timerInfo.IsPastDue == true.

**Comment by paulbatum** [comment]
The fact that this happens for past due timers is particularly problematic in a scenario where timers are being used to regularly register the function app as a webhook endpoint with some other system, because that system is likely to make a call back to the function app to validate the webhook endpoint, and that call will fail because the functions host is not accepting requests. It only takes a small delay in cold start to cause the past-due condition to be true.

I made a simple repro for the behavior:
```csharp
    public static class Functions
    {
        [FunctionName("Timer")]
        public static async Task RunTimer(
            [TimerTrigger("0 * * * * *")] TimerInfo myTimer, 
            ILogger log)            
        {            
            log.LogInformation($"Timer function started running...");
            
            var hostname = Environment.GetEnvironmentVariable("WEBSITE_HOSTNAME") ?? "localhost:7071";
            var protocol = $"http{(hostname.Contains(":") ? "" : "s")}";
            var endpoint = $"{protocol}://{hostname}/api/foo";
            log.LogInformation($"Using endpoint: {endpoint}");

            var watch = Stopwatch.StartNew();

            try
            {                
                var client = new HttpClient();
                var result = await client.GetAsync(endpoint);
                log.LogInformation($"Got response {result.StatusCode} after {watch.ElapsedMilliseconds}ms");
            }            
            finally
            {
                watch.Stop();
            }
        }

        [FunctionName("Foo")]
        public static async Task<IActionResult> RunFoo(
            [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post", Route = null)] HttpRequest req,
            ILogger log)
        {
            log.LogInformation("Running foo...");
            await Task.Yield();
            return new OkObjectResult("foo");
        }
    }
```

Output when running locally is below - first execution fails because of the past due condition, second invocation succeeds.

```
Azure Functions Core Tools
Core Tools Version:       3.0.3442 Commit hash: 6bfab24b2743f8421475d996402c398d2fe4a9e0  (64-bit)
Function Runtime Version: 3.0.15417.0

[2021-05-01T00:02:08.176Z] Found D:\code\temp\TimerLoopbackRequest\TimerLoopbackRequest\TimerLoopbackRequest.csproj. Using for user secrets file configuration.

Functions:

        Foo: [GET,POST] http://localhost:7071/api/Foo

        Timer: timerTrigger

For detailed output, run func with --verbose flag.
[2021-05-01T00:02:10.757Z] Executing 'Timer' (Reason='Timer fired at 2021-04-30T17:02:10.7275781-07:00', Id=d30dfb4c-a1dc-487d-b48e-b4e6027e8adb)
[2021-05-01T00:02:10.758Z] Trigger Details: UnscheduledInvocationReason: IsPastDue, OriginalSchedule: 2021-04-30T17:02:00.0000000-07:00
[2021-05-01T00:02:10.766Z] Timer function started running...
[2021-05-01T00:02:10.767Z] Using endpoint: http://localhost:7071/api/foo
[2021-05-01T00:02:15.215Z] Executed 'Timer' (Failed, Id=d30dfb4c-a1dc-487d-b48e-b4e6027e8adb, Duration=4477ms)
[2021-05-01T00:02:15.217Z] System.Private.CoreLib: Exception while executing function: Timer. System.Net.Http: No connection could be made because the target machine actively refused it. System.Private.CoreLib: No connection could be made because the target machine actively refused it.
[2021-05-01T00:02:20.506Z] Host lock lease acquired by instance ID '00000000000000000000000004884EE9'.
[2021-05-01T00:03:00.010Z] Executing 'Timer' (Reason='Timer fired at 2021-04-30T17:03:00.0088257-07:00', Id=ffbf53b5-29d8-43a2-af58-58709d9c8ba5)
[2021-05-01T00:03:00.013Z] Timer function started running...
[2021-05-01T00:03:00.015Z] Using endpoint: http://localhost:7071/api/foo
[2021-05-01T00:03:02.329Z] Executing 'Foo' (Reason='This function was programmatically called via the host APIs.', Id=e78b54cb-5d29-4252-ab20-524d1556cff6)
[2021-05-01T00:03:02.338Z] Running foo...
[2021-05-01T00:03:02.343Z] Executed 'Foo' (Succeeded, Id=e78b54cb-5d29-4252-ab20-524d1556cff6, Duration=16ms)
[2021-05-01T00:03:02.388Z] Got response OK after 2370ms
[2021-05-01T00:03:02.390Z] Executed 'Timer' (Succeeded, Id=ffbf53b5-29d8-43a2-af58-58709d9c8ba5, Duration=2381ms)
```

**Comment by paulbatum** [comment]
@mathewc This behavior is so counter-intuitive, it costs a bunch of time to figure out what is going on every time it comes up. Would it be a breaking change to fix this?

**Comment by mathewc** [comment]
No I don't think it would be breaking. Rather than starting the invocation in StartAsync, we could consider setting a flag so the first interval used after startup is zero. I.e. [here](https://github.com/Azure/azure-webjobs-sdk-extensions/blob/3d7b2b2bc75927f3e84ed231f14c175ff342ef23/src/WebJobs.Extensions/Extensions/Timers/Listener/TimerListener.cs#L279) rather than computing the next schedule interval, we use a short/zero interval. This would allow the host listener startup to complete, and the invocation would run after that. Would have to experiment with your repro to make sure it would still give enough time for the host http trigger listener to register fully before your timer self-request runs. However it should work, because there's middleware that delays any incoming http requests until the host has started up sufficiently to handle http requests.

**Comment by fabiocav** [comment]
Bumping this up as a result of another internal thread. @mathewc should we assign t his work to the next sprint? This behavior also has an odd impact on OOP workers.

**Comment by mathewc** [comment]
Yes, let's schedule it.

**Comment by alrod** [comment]
Related CRI:
https://portal.microsofticm.com/imp/v3/incidents/details/249734654/home

**Comment by fabiocav** [comment]
Working on assignment for this. Still in 106, but it's currently a stretch goal.

**Comment by fabiocav** [comment]
Not assigned in 106. Moving this to sprint 107

**Comment by mathewc** [comment]
Fix out for PR for this: https://github.com/Azure/azure-webjobs-sdk-extensions/pull/737

**Comment by fabiocav** [comment]
Fixed by Azure/azure-webjobs-sdk-extensions#737
